File: src/dotgen.rs
use crate::{ast::*, do_, monad::Monad};

static mut COUNTER: u32 = 0;

unsafe fn inc() -> u32 {
    unsafe {
        COUNTER += 1;
        COUNTER
    }
}

#[derive(Debug, Clone)]
struct NodeIndex {
    name: Option<String>,
    idx: u32,
}

impl NodeIndex {
    fn new(name: Option<String>) -> Self {
        NodeIndex {
            name,
            idx: unsafe { inc() },
        }
    }
}

struct Writer<T> {
    value: T,
    output: String,
}

impl<T> Monad<T> for Writer<T> {
    type Output<U> = Writer<U>;

    fn bind<U>(self, f: impl FnOnce(T) -> Writer<U>) -> Writer<U> {
        let Writer { value, output } = f(self.value);
        Writer {
            value,
            output: self.output + &output,
        }
    }
}

pub fn to_dot(ast: &Expr, name: Option<String>) -> String {
    match name {
        Some(name) => {
            let root = NodeIndex::new(Some(name.clone()));
            format!(
                "subgraph {} {{\n\t{} [shape=point, width=0.1];\n{}\n}}",
                name,
                root.clone().idx,
                ast.to_graph(root).output
            )
        }
        None => {
            let root = NodeIndex::new(None);
            format!(
                "digraph {{\n\t{} [shape=point, width=0.1];\n{}\n}}",
                root.clone().idx,
                ast.to_graph(root).output
            )
        }
    }
}

trait ToGraph {
    fn to_graph(&self, parent: NodeIndex) -> Writer<NodeIndex>;
}

fn new_node<T>(name: T, parent: NodeIndex, color: &str) -> Writer<NodeIndex>
where
    T: ToString,
{
    let cur = NodeIndex::new(parent.name.clone());
    Writer {
        value: cur.clone(),
        output: format!(
            "\t{} [shape=point, width=0.1, color=\"{}\"];\n\t{} -> {} [label=\"{}\", arrowhead=none, color=\"{}\", fontcolor=\"{}\"];\n",
            cur.clone().idx,
            color,
            parent.idx,
            cur.idx,
            name.to_string(),
            color,
            color,
        ),
    }
}

impl ToGraph for Variable {
    fn to_graph(&self, parent: NodeIndex) -> Writer<NodeIndex> {
        new_node(self.0.clone(), parent, "black")
    }
}

impl ToGraph for Expr {
    fn to_graph(&self, parent: NodeIndex) -> Writer<NodeIndex> {
        match self {
            Expr::Var(_) | Expr::Num(_) | Expr::True | Expr::False | Expr::Unit => {
                new_node(self, parent, "red")
            }
            Expr::Addop { binop, left, right } => do_!(
                new_node(binop, parent, "red") => cur,
                left.to_graph(cur.clone()),
                right.to_graph(cur)
            ),
            Expr::Mulop { binop, left, right } => do_!(
                new_node(binop, parent, "red") => cur,
                left.to_graph(cur.clone()),
                right.to_graph(cur)
            ),
            Expr::If { cond, then_, else_ } => do_!(
                new_node("if", parent, "red") => cur,
                cond.to_graph(cur.clone()),
                then_.to_graph(cur.clone()),
                else_.to_graph(cur)
            ),
            Expr::Relop { left, right, relop } => do_!(
                new_node(relop, parent, "red") => cur,
                left.to_graph(cur.clone()),
                right.to_graph(cur)
            ),
            Expr::And { left, right } => do_!(
                new_node("&&", parent, "red") => cur,
                left.to_graph(cur.clone()),
                right.to_graph(cur)
            ),
            Expr::Or { left, right } => do_!(
                new_node("||", parent, "red") => cur,
                left.to_graph(cur.clone()),
                right.to_graph(cur)
            ),
            Expr::Pair { left, right } => do_!(
                new_node("pair", parent, "red") => cur,
                left.to_graph(cur.clone()),
                right.to_graph(cur)
            ),
            Expr::App { lam, arg } => do_!(
                new_node("app", parent, "red") => cur,
                lam.to_graph(cur.clone()),
                arg.to_graph(cur)
            ),
            Expr::Lam { x, tau, e } => do_!(
                new_node("λ", parent, "red") => cur,
                x.to_graph(cur.clone()),
                tau.to_graph(cur.clone()),
                e.to_graph(cur)
            ),
            Expr::Fix { x, tau, e } => do_!(
                new_node("fix", parent, "red") => cur,
                x.to_graph(cur.clone()),
                tau.to_graph(cur.clone()),
                e.to_graph(cur)
            ),
            Expr::Project { e, d } => do_!(
                new_node(match d {
                    Direction::Left => "P_left",
                    Direction::Right => "P_right",
                }, parent, "red") => cur,
                e.to_graph(cur)
            ),
            Expr::Inject { e, d, tau } => do_!(
                new_node(match d {
                    Direction::Left => "I_left",
                    Direction::Right => "I_right",
                }, parent, "red") => cur,
                e.to_graph(cur.clone()),
                tau.to_graph(cur)
            ),
            Expr::Case {
                e,
                xleft,
                eleft,
                xright,
                eright,
            } => do_!(
                new_node("case", parent, "red") => cur,
                e.to_graph(cur.clone()),
                xleft.to_graph(cur.clone()),
                eleft.to_graph(cur.clone()),
                xright.to_graph(cur.clone()),
                eright.to_graph(cur)
            ),
            Expr::TyApp { e, tau } => do_!(
                new_node("tyapp", parent, "red") => cur,
                e.to_graph(cur.clone()),
                tau.to_graph(cur)
            ),
            Expr::TyLam { a, e } => do_!(
                new_node("Λ", parent, "red") => cur,
                a.to_graph(cur.clone()),
                e.to_graph(cur)
            ),
            Expr::Fold { e, tau } => do_!(
                new_node("fold", parent, "red") => cur,
                e.to_graph(cur.clone()),
                tau.to_graph(cur)
            ),
            Expr::Unfold(e) => do_!(
                new_node("unfold", parent, "red") => cur,
                e.to_graph(cur)
            ),
            Expr::Import {
                x,
                a,
                e_mod,
                e_body,
            } => do_!(
                new_node("import", parent, "red") => cur,
                x.to_graph(cur.clone()),
                a.to_graph(cur.clone()),
                e_mod.to_graph(cur.clone()),
                e_body.to_graph(cur)
            ),
            Expr::Export {
                e,
                tau_adt,
                tau_mod,
            } => do_!(
                new_node("export", parent, "red") => cur,
                e.to_graph(cur.clone()),
                tau_adt.to_graph(cur.clone()),
                tau_mod.to_graph(cur)
            ),
        }
    }
}

impl ToGraph for Type {
    fn to_graph(&self, parent: NodeIndex) -> Writer<NodeIndex> {
        match self {
            Type::Num | Type::Bool | Type::Unit | Type::Var(_) => new_node(self, parent, "blue"),
            Type::Product { left, right } => do_!(
                new_node("*", parent, "blue") => cur,
                left.to_graph(cur.clone()),
                right.to_graph(cur)
            ),
            Type::Sum { left, right } => do_!(
                new_node("+", parent, "blue") => cur,
                left.to_graph(cur.clone()),
                right.to_graph(cur)
            ),
            Type::Fn { arg, ret } => do_!(
                new_node("→", parent, "blue") => cur,
                arg.to_graph(cur.clone()),
                ret.to_graph(cur)
            ),
            Type::Rec { a, tau } => do_!(
                new_node("rec", parent, "blue") => cur,
                a.to_graph(cur.clone()),
                tau.to_graph(cur)
            ),
            Type::Forall { a, tau } => do_!(
                new_node("∀", parent, "blue") => cur,
                a.to_graph(cur.clone()),
                tau.to_graph(cur)
            ),
            Type::Exists { a, tau } => do_!(
                new_node("∃", parent, "blue") => cur,
                a.to_graph(cur.clone()),
                tau.to_graph(cur)
            ),
        }
    }
}

File: src/DotGen.hs
module DotGen where

import AST
import ASTUtil (Symbol(..))

data NodeIndex = NodeIndex
    { name :: Maybe String
    , idx :: Int
    }

newNodeIndex :: Maybe String -> Int -> NodeIndex
newNodeIndex n i = NodeIndex {name = n, idx = i}

data Writer a = Writer
    { value :: a
    , output :: String
    }

instance Functor Writer where
    fmap f (Writer v o) = Writer (f v) o

instance Applicative Writer where
    pure a = Writer a ""
    Writer f o1 <*> Writer v o2 = Writer (f v) (o1 ++ o2)

instance Monad Writer where
    return = pure
    (Writer v o1) >>= f = let Writer v' o2 = f v in Writer v' (o1 ++ o2)

tell :: String -> Writer ()
tell = Writer ()

newNode :: Show a => a -> NodeIndex -> Int -> String -> Writer (NodeIndex, Int)
newNode v parent nextIdx color = do
    let cur = newNodeIndex (name parent) nextIdx
    tell $ "\t" ++ show (idx cur) ++ " [shape=point, width=0.1, color=\"" ++ color ++ "\"];\n" ++
           "\t" ++ show (idx parent) ++ " -> " ++ show (idx cur) ++
           " [label=" ++ show v ++ ", arrowhead=none, color=\"" ++
           color ++ "\", fontcolor=\"" ++ color ++ "\"];\n"
    return (cur, nextIdx + 1)

class ToGraph a where
    toGraph :: a -> NodeIndex -> Int -> Writer Int

instance ToGraph Variable where
    toGraph v parent counter = do
        (_, next) <- newNode (show v) parent counter "black"
        return next

instance ToGraph Expr where
    toGraph e parent counter = case e of
        EVar _ -> do
            (_, next) <- newNode (show e) parent counter "red"
            return next
        EDeBruijn _ -> do
            (_, next) <- newNode (show e) parent counter "red"
            return next
        ENum _ -> do
            (_, next) <- newNode (show e) parent counter "red"
            return next
        ETrue -> do
            (_, next) <- newNode (show e) parent counter "red"
            return next
        EFalse -> do
            (_, next) <- newNode (show e) parent counter "red"
            return next
        EUnit -> do
            (_, next) <- newNode (show e) parent counter "red"
            return next

        EAddop op left right -> do
            (cur, next1) <- newNode (show op) parent counter "red"
            next2 <- toGraph left cur next1
            toGraph right cur next2

        EMulop op left right -> do
            (cur, next1) <- newNode (show op) parent counter "red"
            next2 <- toGraph left cur next1
            toGraph right cur next2

        EIf cond then_ else_ -> do
            (cur, next1) <- newNode "if" parent counter "red"
            next2 <- toGraph cond cur next1
            next3 <- toGraph then_ cur next2
            toGraph else_ cur next3

        ERelop op left right -> do
            (cur, next1) <- newNode (show op) parent counter "red"
            next2 <- toGraph left cur next1
            toGraph right cur next2

        EAnd left right -> do
            (cur, next1) <- newNode "&&" parent counter "red"
            next2 <- toGraph left cur next1
            toGraph right cur next2

        EOr left right -> do
            (cur, next1) <- newNode "||" parent counter "red"
            next2 <- toGraph left cur next1
            toGraph right cur next2

        EPair left right -> do
            (cur, next1) <- newNode "pair" parent counter "red"
            next2 <- toGraph left cur next1
            toGraph right cur next2

        EApp lam arg -> do
            (cur, next1) <- newNode "app" parent counter "red"
            next2 <- toGraph lam cur next1
            toGraph arg cur next2

        ELam x _ e -> do
            (cur, next1) <- newNode "λ" parent counter "red"
            next2 <- toGraph x cur next1
            toGraph e cur next2

        EFix x e -> do
            (cur, next1) <- newNode "fix" parent counter "red"
            next2 <- toGraph x cur next1
            toGraph e cur next2

        EProject e d -> do
            (cur, next1) <- newNode (case d of L -> "P_left"; R -> "P_right") parent counter "red"
            toGraph e cur next1

        EInject e d -> do
            (cur, next1) <- newNode (case d of L -> "I_left"; R -> "I_right") parent counter "red"
            toGraph e cur next1

        ECase e xleft eleft xright eright -> do
            (cur, next1) <- newNode "case" parent counter "red"
            next2 <- toGraph e cur next1
            next3 <- toGraph xleft cur next2
            next4 <- toGraph eleft cur next3
            next5 <- toGraph xright cur next4
            toGraph eright cur next5

        ELet x _ e_x e_in -> do
            (cur, next1) <- newNode "let" parent counter "red"
            next2 <- toGraph x cur next1
            next3 <- toGraph e_x cur next2
            toGraph e_in cur next3

instance ToGraph Type where
    toGraph t parent counter = case t of
        TNum -> do
            (_, next) <- newNode (show t) parent counter "blue"
            return next
        TBool -> do
            (_, next) <- newNode (show t) parent counter "blue"
            return next
        TUnit -> do
            (_, next) <- newNode (show t) parent counter "blue"
            return next
        TVar _ -> do
            (_, next) <- newNode (show t) parent counter "blue"
            return next

        TProduct left right -> do
            (cur, next1) <- newNode "*" parent counter "blue"
            next2 <- toGraph left cur next1
            toGraph right cur next2

        TSum left right -> do
            (cur, next1) <- newNode "+" parent counter "blue"
            next2 <- toGraph left cur next1
            toGraph right cur next2

        TFn arg ret -> do
            (cur, next1) <- newNode "→" parent counter "blue"
            next2 <- toGraph arg cur next1
            toGraph ret cur next2

        TForall a tau -> do
            (cur, next1) <- newNode "∀" parent counter "blue"
            next2 <- toGraph a cur next1
            toGraph tau cur next2

toDot :: Expr -> Maybe String -> String
toDot ast name = case name of
    Just n ->
        let root = newNodeIndex (Just n) 0
            result = toGraph (toDebruijn ast) root 1
        in "subgraph " ++ n ++ " {\n\t" ++ show (idx root) ++
           " [shape=point, width=0.1];\n" ++ output result ++ "}"
    Nothing ->
        let root = newNodeIndex Nothing 0
            result = toGraph (toDebruijn ast) root 1
        in "digraph {\n\t" ++ show (idx root) ++
           " [shape=point, width=0.1];\n" ++ output result ++ "}"
